src/librustc/infer/error_reporting/anon_anon_conflict.rs

```diff
@@ -54,30 +54,57 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         let scope_def_id_sub = anon_reg_sub.def_id;
         let bregion_sub = anon_reg_sub.boundregion;
 
-        let ty_sup = or_false!(self.find_anon_type(sup, &bregion_sup));
+        let anon_ty_sup = or_false!(self.find_anon_type(sup, &bregion_sup));
 
-        let ty_sub = or_false!(self.find_anon_type(sub, &bregion_sub));
+        let anon_ty_sub = or_false!(self.find_anon_type(sub, &bregion_sub));
+        let (ty_sup, ty_fndecl_sup) = anon_ty_sup;
+        let (ty_sub, ty_fndecl_sub) = anon_ty_sub;
 
-        let (main_label, label1, label2) = if let (Some(sup_arg), Some(sub_arg)) =
+        let (spanfin1, spanfin2, main_label, label) = if let (Some(sup_arg),
+                                                                       Some(sub_arg)) =
             (self.find_arg_with_anonymous_region(sup, sup),
              self.find_arg_with_anonymous_region(sub, sub)) {
 
-            let (anon_arg_sup, is_first_sup, anon_arg_sub, is_first_sub) =
-                (sup_arg.arg, sup_arg.is_first, sub_arg.arg, sub_arg.is_first);
-            if self.is_self_anon(is_first_sup, scope_def_id_sup) ||
-               self.is_self_anon(is_first_sub, scope_def_id_sub) {
-                return false;
-            }
-
-            if self.is_return_type_anon(scope_def_id_sup, bregion_sup) ||
-               self.is_return_type_anon(scope_def_id_sub, bregion_sub) {
-                return false;
-            }
+            let (anon_arg_sup, anon_arg_sub) = (sup_arg.arg, sub_arg.arg);
 
             if anon_arg_sup == anon_arg_sub {
-                (format!("this type was declared with multiple lifetimes..."),
-                 format!(" with one lifetime"),
-                 format!(" into the other"))
+                let (span_1, span_2, span_label) =
+                    if self.is_return_type_anon(scope_def_id_sup, bregion_sup, ty_fndecl_sup)
```

Comment:
Hmm, this logic is sort of complex. First off, let's store `self.is_return_type_anon(scope_def_id_sup, bregion_sup, ty_fndecl_sup)` into a variable:

```rust
let sup_is_ret_type = self.is_return_type_anon(scope_def_id_sup, bregion_sup, ty_fndecl_sup);
```

and same for `sub`. Then I think I would structure the remaining logic something like this:

```rust
let sup_is_ret_type = self.is_return_type_anon(scope_def_id_sup, bregion_sup, ty_fndecl_sup);
let sub_is_ret_type = self.is_return_type_anon(scope_def_id_sub, bregion_sub, ty_fndecl_sub);
let (span_1, span_2, span_label) = match (sup_is_ret_type, sub_is_ret_type) {
    (None, None) =>
        (ty_sup.span, ty_sub.span, format!("this type was declared with multiple lifetimes...")),
    (Some(ret_span), _) =>
        (ty_sub.span, ret_span,  format!("this parameter and the return type are declared with different lifetimes...")),
    (_, Some(ret_span)) =>
        (ty_sup.span, ret_span,  format!("this parameter and the return type are declared with different lifetimes...")),
};
```

However, this raises the question -- what if both sub and sup appear in the return type? The existing code seems to ignore this case, but it's probably worth adding something like this:

```rust
(Some(ret_span1), Some(ret_span2)) =>
    (ret_span1, ret_span2, format!("the return type is declared with different lifetimes...")),
```

That said, maybe better to wait until we have a test for this case, since I'm not sure off-hand how it can really arise.